### ISSUE TITLE
libsm64's usage as an external dependency

### ISSUE THREAD

**Issue by jgcodes2020** [branpk/wafel#23]
It seems that I've hit a roadblock in libsm64-utils: I've hit the inevitable struct-padding issue. Since the `MarioState` struct isn't manually padded for 64-bit systems, the offsets won't be right. So I'm not sure how I'm supposed to provide integration with the library.

EDIT: To be more clear, how did you access struct members in a compiler-independent manner?

**Comment by branpk** [comment]
Struct offsets are included in the DWARF debugging info. When wafel was python-only, I used pefile for reading the DLL, and pyelftools for parsing the DWARF info. [This code](https://github.com/branpk/wafel/blob/84291ad6cf4f42b7a14af4129be381db341d17f8/wafel/core/dll_data_spec.py) might be helpful (feel free to copy+paste).

Depending on your use case, you could also just generate a json file with all struct offsets using wafel's current code for 32 and 64 bit and include that in your library. I can help if you want to go this route.

**Comment by jgcodes2020** [comment]
I've searched for PE parsing and DWARF debugging libraries, and I've hit a dead end for those too (library is in C++), looks like I'm going to need a json file

EDIT: What version of DWARF does wafel use?

**Comment by jgcodes2020** [comment]
I might need some kind of Rust project to

**Comment by branpk** [comment]
I don't know what version of DWARF is used off the top of my head, but I'd guess 5. I suppose it would depend on what version of gcc you use to compile libsm64.

I think the best way forward here is for me to extract wafel's [current code](https://github.com/branpk/wafel/blob/master/wafel_core/src/dll/layout.rs) into a crate. You'd have to figure out how to use it from C/C++, but this way it could be used in other projects and you wouldn't have to worry about updating the json with new decomp changes.

Though in the short term, pulling a json file would be quickest.

**Comment by jgcodes2020** [comment]
Never mind: I've found some libraries that will do the trick for debug parsing.

EDIT: What sections of the DLL have the debug info in them?

**Comment by branpk** [comment]
I've extracted out the code that Wafel uses for this into its own crate [wafel_layout](https://github.com/branpk/wafel/tree/master/wafel_layout).

Probably more relevant for you, I also added a new executable target `libsm64_layout` to Wafel which takes the libsm64 dll as input, and produces a json file with types, struct offsets, etc. To build it, run `cargo build --release --bin libsm64_layout` (only requires Rust). It outputs to `target/release/libsm64_layout.exe`.
